Common VCFWriter and honor index/md5 arguments.

cmnbroad opened the pull request:
Common VCFWriter creation, and fix for https://github.com/broadinstitute/gatk/issues/365 (for non-Spark tools).

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java:
```diff
@@ -43,12 +48,31 @@
+    @Argument(fullName=StandardArgumentDefinitions.CREATE_OUTPUT_BAM_INDEX_LONG_NAME,
+            shortName=StandardArgumentDefinitions.CREATE_OUTPUT_BAM_INDEX_SHORT_NAME,
+            doc = "If true, create a BAM/CRAM index when writing a coordinate-sorted BAM/CRAM file.", optional=true)
     public boolean createOutputBamIndex = true;
```
in general we avoid boolean flags that default to true.  Could you invert these so they're `dontCreateOutputBamIndex` etc.

droazen replied:
I disagree in this case, actually (though I agree with the general principle of not having boolean args that default to true, in general). If @cmnbroad inverted this arg, it would be inconsistent with `createOutputBamMD5`, etc., below, which would be confusing. Recommend keeping all of these args as `create*` instead of a mix of `dontCreate*` and `create*`

cmnbroad replied:
Leaving these as is.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java:
```diff
@@ -421,6 +445,47 @@ public final SAMFileGATKReadWriter createSAMWriter(final File outputFile, final
+                logger.warn("An variant index will not be created - a sequence dictionary is required to create an output index");
+                // fall through and create without an index
+                options = lenientVCFProcessing ?
+                        new Options[]{Options.ALLOW_MISSING_FIELDS_IN_HEADER} :
```
This logic is way more complicated than it needs to be.  Can you just use a 'List' instead of the '[]' and simply check each case and append or don't instead of repeating the ternary expressions for case?  Then just convert to array at the end.

cmnbroad replied:
Yeah. Much better. Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java:
```diff
@@ -421,6 +445,47 @@ public final SAMFileGATKReadWriter createSAMWriter(final File outputFile, final
+     * this tool, and default index, leniency and md5 generation settings.
+     *
+     * @param outFile output File for this writer. May not be null.
+     * @returns VariantContextWriter must be closed by the caller
```
Why not have GatkTool keep a list of these and close them itself?

lbergelson replied:
I guess the existing createSAMWriter() doesn't do it either, but it seems like maybe both of them should keep track of what they've created and try to close them at the end...

droazen replied:
The current contract is that walkers must close their own writers. We can change this (and probably should), but I think it should be done in a separate ticket.

cmnbroad replied:
We have a ticket (https://github.com/broadinstitute/gatk/issues/121) for the work to create and manage these automatically.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -40,40 +42,69 @@ public static boolean isInformative(final double[] gls) {
     }
 
+    // Determine the variant file type from the file extension.
+    // This should really be exposed in htsjdk from AbstractFeatureReader
```
could you open a ticket in htsjdk so we don't forget about it?

cmnbroad replied:
I'm reconsidering whether I should be doing this at all, see comment above.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/GATKToolUnitTest.java:
```diff
@@ -281,4 +321,207 @@ public void testBestSequenceDictionary_fromNothing() throws Exception {
+        final File outputFile = new File(tmpDir.getAbsolutePath(), "createVCFTest" + outputExtension);
+
+        List<String> args = new ArrayList<>();
+        args.add("--input"); args.add(inputFile.getCanonicalPath());
```
we have some ArgumentsBuilder convenience methods to do this less verbosely if you prefer

cmnbroad replied:
Done.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/engine/GATKToolUnitTest.java:
```diff
@@ -281,4 +321,207 @@ public void testBestSequenceDictionary_fromNothing() throws Exception {
+        }
+
+        final CommandLineParser clp = new CommandLineParser(tool);
+        clp.parseArguments(System.out, args.toArray(new String[args.size()]));
```
args.getArgsArray()

cmnbroad replied:
Done.

lbergelson commented:
@cmnbroad Back you with a minor refactoring request.  It seems to me that GATKTool should try and close it's own writers, but maybe there's a reason not to that I'm missing?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java:
```diff
@@ -421,6 +445,47 @@ public final SAMFileGATKReadWriter createSAMWriter(final File outputFile, final
     }
 
     /**
+     * Creates a VariantContextWriter who's outputFile type is determined by
```
`who's`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/filters/VariantFiltration.java:
```diff
@@ -199,8 +198,7 @@ private String possiblyInvertFilterExpression( final String description ){
     private void initializeVcfWriter() {
-        //TODO remove hardwiring to output VCFs
-        writer = new VariantContextWriterBuilder().setOutputFile(out).setOutputFileType(VariantContextWriterBuilder.OutputType.VCF).unsetOption(Options.INDEX_ON_THE_FLY).build();
+        writer = createVCFWriter(new File(out));
```
Can you have a look at `HaplotypeCallerEngine.makeVCFWriter()` in gatk-protected, and see whether that method can be made to be consistent with your work in this branch, and create a new ticket to patch it? It calls `GATKVariantContextUtils.createVCFWriter()`, but sometimes it constructs a new `GVCFWriter`. You may need a way to create a `GVCFWriter` at the `GATKTool` level, as well as a static entry point not tied to the tool instance (but we can deal with this in a separate ticket)

cmnbroad replied:
Created a ticket https://github.com/broadinstitute/gatk/issues/2102.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -40,40 +42,69 @@ public static boolean isInformative(final double[] gls) {
 
     /**
-     * Creates a VariantContextWriter who's outputFile type is <code>VariantContextWriterBuilder.OutputType.VCF</code>.
+     * Creates a VariantContextWriter who's outputFile type is based on the extension of the output file name.
```
`who's` -> `whose`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -40,40 +42,69 @@ public static boolean isInformative(final double[] gls) {
+            logger.warn(String.format(
+                    "Can't determine output variant file format from output file extension \"%s\". Defaulting to VCF.",
+                    FilenameUtils.getExtension(outFile.getPath())));
+            vcWriterBuilder = vcWriterBuilder.setOutputFileType(VariantContextWriterBuilder.OutputType.VCF);
```
Does this not need to be explicitly set in the other case (`VariantContextWriterBuilder.OutputType.UNSPECIFIED != getVariantFileTypeFromExtension(outFile)`)?

cmnbroad replied:
No, VariantContextWriterBuilder will derive the output type itself based on the file extension, its just that its not very forgiving and throws if it doesn't recognize they extension as a known type (i.e, a name that ends in .tmp). Perhaps I shouldn't be trying to do this at all, and we should just let it throw if such cases. Let me know what you think.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -40,40 +42,69 @@ public static boolean isInformative(final double[] gls) {
+    private static VariantContextWriterBuilder.OutputType getVariantFileTypeFromExtension(final File outputFile) {
+        final String extension = FilenameUtils.getExtension(outputFile.getPath()).toLowerCase();
+
+        if (extension.equals("vcf")) {
```
What about the gzipped versions of these?

droazen replied:
Declare these extensions as constants in `VcfUtils`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -40,40 +42,69 @@ public static boolean isInformative(final double[] gls) {
+     * @param options variable length list of additional Options to be set for this writer
      * @returns VariantContextWriter must be closed by the caller
      */
     public static VariantContextWriter createVCFWriter(
```
This method should be able to create a `GVCFWriter` as well. GATK3 uses the semi-unofficial file extension `.g.vcf` for GVCFs.

lbergelson replied:
Is that something we want this method to do?  We'd have to add additional parameters to this method to take the gqPartitions and default ploidy.  I'm not sure any tools that aren't haplotype caller would want to write a gvcf, so I think it would be ok if it just does the wrapping of the output writer manually still?

cmnbroad replied:
Added this note to the #2102 ticket

cmnbroad replied:
Yeah - I suggested exactly the same thing to @droazen yesterday. It makes more sense that overloading this one.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/GATKToolUnitTest.java:
```diff
@@ -281,4 +321,207 @@ public void testBestSequenceDictionary_fromNothing() throws Exception {
+
+        Assert.assertTrue(outputFile.exists());
+        Assert.assertEquals(outFileIndex.exists(), createIndex);
+        Assert.assertEquals(outFileMD5.exists(), createMD5);
```
Add error messages for when the asserts fail

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -1733,30 +1737,197 @@ public void testTotalPloidy(final int[] ploidies, final int defaultPloidy, final
+                makeSimpleSequenceDictionary(),
+                createMD5,
+                options)) {
+            writeHeaderAndBadVariant(writer, false);
```
Why are you writing a "bad" variant in this `testCreateVCFWriterWithOptions` test case?

cmnbroad replied:
I'm not, its just a bad name. The second arg ("false" in this case) determines whether to write the bad variant. I've separated this into two separate methods to make it clearer.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -1733,30 +1737,197 @@ public void testTotalPloidy(final int[] ploidies, final int defaultPloidy, final
+            throw new IllegalArgumentException("Unknown file extension in createVCFWriter test validation");
+        }
+
+        Assert.assertNotEquals(null, afr);
```
`Assert.assertNotNull()`

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -1733,30 +1737,197 @@ public void testTotalPloidy(final int[] ploidies, final int defaultPloidy, final
+
+    // just make sure we can read the file with the corresponding codec
+    private void verifyFileType(final File resultVCFFile, final String outputExtension, final boolean createIndex) {
+        AbstractFeatureReader<?, ?> afr = null;
```
Close this reader before the method returns

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -1733,30 +1737,197 @@ public void testTotalPloidy(final int[] ploidies, final int defaultPloidy, final
+                    resultVCFFile.getAbsolutePath(),
+                    new VCFCodec(),
+                    createIndex);
+        }
```
Would be simpler to just construct a `FeatureDataSource` and make sure that you don't get an exception, instead of building the reader manually like this.

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java:
```diff
@@ -1733,30 +1737,197 @@ public void testTotalPloidy(final int[] ploidies, final int defaultPloidy, final
+        writer.add(vc);
     }
+
+    public SAMSequenceDictionary makeSimpleSequenceDictionary() {
```
`private`

cmnbroad replied:
Done.

droazen commented:
Did a review pass on this as well, and added my comments. Back to @cmnbroad

cmnbroad commented:
I've addressed all the code review comments. I'm questioning the value of setting the VCF writer output type by inspecting the output file extension though. I'm wondering if I should just remove this and let htsjdk throw it can't determine the type from the extension.

lbergelson commented:
@cmnbroad I think the defaulting to vcf if the file extension is ambiguous is probably better behavior for gatk than crashing.   We can read a vcf with no extension, so it seems symmetrical at least.  What's the downside of including it, is it just the fairly minor added complexity?

cmnbroad commented:
@lbergelson Yes, I was thinking we're just making it easy to do something suboptimal (make file with a random extension). The complexity is pretty minimal - I'll leave it in since it sounds like you think its at least somewhat useful.

lbergelson commented:
@cmnbroad 👍  Rebase and merge when ready.  I think people expect to be able to write files with any extension they feel like, and the risks are pretty minimal.
